atc/db/pipeline_lifecycle_test.go

```diff
@@ -10,70 +12,124 @@ import (
-				childPipeline.Reload()
-				Expect(childPipeline.Archived()).To(BeTrue())
-			})
-		})
+			Context("parent pipeline is destroyed", func() {
+				BeforeEach(func() {
+					defaultPipeline.Destroy()
+				})
 
-		Context("parent pipeline is archived", func() {
-			BeforeEach(func() {
-				defaultPipeline.Archive()
+				It("should archive all child pipelines", func() {
+					childPipeline.Reload()
+					Expect(childPipeline.Archived()).To(BeTrue())
+				})
 			})
 
-			It("should archive all child pipelines", func() {
-				childPipeline.Reload()
-				Expect(childPipeline.Archived()).To(BeTrue())
+			Context("parent pipeline is archived", func() {
+				BeforeEach(func() {
+					defaultPipeline.Archive()
+				})
+
+				It("should archive all child pipelines", func() {
+					childPipeline.Reload()
+					Expect(childPipeline.Archived()).To(BeTrue())
+				})
 			})
-		})
 
-		Context("job is removed from the parent pipeline", func() {
-			BeforeEach(func() {
-				defaultPipelineConfig.Jobs = atc.JobConfigs{
-					{
-						Name: "a-different-job",
-					},
-				}
-				defaultTeam.SavePipeline("default-pipeline", defaultPipelineConfig, defaultPipeline.ConfigVersion(), false)
+			Context("job is removed from the parent pipeline", func() {
+				BeforeEach(func() {
+					defaultPipelineConfig.Jobs = atc.JobConfigs{
+						{
+							Name: "a-different-job",
+						},
+					}
+					defaultTeam.SavePipeline("default-pipeline", defaultPipelineConfig, defaultPipeline.ConfigVersion(), false)
+				})
+
+				It("archives all child pipelines set by the deleted job", func() {
+					childPipeline.Reload()
+					Expect(childPipeline.Archived()).To(BeTrue())
+				})
 			})
+		})
 
-			It("archives all child pipelines set by the deleted job", func() {
-				childPipeline.Reload()
-				Expect(childPipeline.Archived()).To(BeTrue())
+		Context("pipeline does not have a parent job and build ID", func() {
+			It("Should not archive the pipeline", func() {
+				defaultPipeline.Reload()
+				Expect(defaultPipeline.Archived()).To(BeFalse())
 			})
 		})
 	})
 
-	Context("pipeline does not have a parent job and build ID", func() {
-		It("Should not archive the pipeline", func() {
-			defaultPipeline.Reload()
-			Expect(defaultPipeline.Archived()).To(BeFalse())
+	Describe("RemoveBuildEventsForDeletedPipelines", func() {
+		var (
+			pipeline1 db.Pipeline
+			pipeline2 db.Pipeline
+		)
+
+		BeforeEach(func() {
+			pipeline1 = defaultPipeline
+			pipeline2, _, _ = defaultTeam.SavePipeline("pipeline2", defaultPipelineConfig, 0, false)
+		})
+
+		pipelineBuildEventsExists := func(id int) bool {
+			var exists bool
+			err = dbConn.QueryRow(fmt.Sprintf("SELECT EXISTS (SELECT 1 FROM information_schema.tables WHERE table_name = 'pipeline_build_events_%d')", id)).Scan(&exists)
+			Expect(err).ToNot(HaveOccurred())
+
+			return exists
+		}
+
+		It("drops the pipeline_build_events_x table for each deleted pipeline", func() {
+			pipeline1.Destroy()
```

Comment:
nit: It's not a big deal but I might suggest to check the error for destroying the pipeline? 🤔  Just to make sure that it didn't error and that is actually what is causing `RemoveBuildEventsForDeletedPipelines` to not remove the build events for that pipeline.

Reply:
I guess what I'm struggling with is - what would cause `pipeline1.Destroy()` to fail? I might be wrong, but I think something would have to be pretty borked with the postgres runner for this to error, in which case we'd likely get a whole bunch of errors in other test cases, anyway.

IMO having the expectations focused on what the test case is actually testing is typically more valuable than erroring early, at least when the error is unlikely to happen/would affect a bunch of other tests. Feel free to challenge me on this - you've worked on this part of the codebase far longer than I have, and it's definitely possible that having more `Expect(err).ToNot(HaveOccurred())`s sprinkled around have made tests more robust/easy to debug for you.

cc @chenbh

Reply:
Going to push changes to add in the error checks since it's pretty minor anyway and I'd ideally like to get this in for 6.5. Found a decent compromise in adding a `destroy` helper that calls `pipeline.Destroy()` and ensures no error occurred. That way, at least the tests can remain easy to read while still giving us more helpful errors if an unexpected error occurs.

Reply:
Hm I don't think I have a very strong opinion on it, I've just seen times within db tests where because there was an error check missing it ended up taking much longer to figure out what went wrong. I think it might make your tests a bit more resilient to refactors because even though right now it is a simple `delete` query, it can be possible that someone in the future adds more logic within `pipeline.Destroy()` where an error can be thrown outside of the `delete` query.  But that being said, I really don't have a strong preference because I can also understand it taking away from the focus of the test. I think it's really up to you haha whatever you think is best! I just wanted to point it out just in case.